### PR TITLE
Don't call methods of @user if parameters are required

### PR DESCRIPTION
**Pull request by valdis**
If attribute methods are overwritten to have parameters
e.g.
```
class User < ApplicationRecord
  def name(formating)
  end
end
```
Airbrake will fail with error `ArgumentError (wrong number of arguments (given 0, expected 1))`
and no error message will be sent to Airbrake

### CODE REVIEW

**Comment by kyrylo** [comment]
Looks good, but it needs:

1. Tests
2. Green build
3. A sip of hot tea/coffee after the job is done 😁

**Comment by valdis** [comment]
Should be good now

File: lib/airbrake/rack/user.rb [hunk 1]
```diff
@@ -50,7 +50,9 @@ def as_json
       private
 
       def try_to_get(key)
-        String(@user.__send__(key)) if @user.respond_to?(key)
+        String(@user.__send__(key)) if @user.respond_to?(key) &&
+                                       (@user.method(key).arity.zero? ||
```
Comment (xymbol) [inline lib/airbrake/rack/user.rb]: Would extract a method here. Then, do:

```ruby
def can_call?(sym)
  @user.respond_to?(sym) && @user.method(sym).arity <= 0
end
```

File: spec/unit/rack/user_spec.rb [hunk 2]
```diff
@@ -189,5 +189,19 @@ def current_user
         expect(user_data).to be_empty
       end
     end
+
+    context "when Rack user field getters require parameters" do
+      class UserRequiringParameters
```
Comment (xymbol) [inline spec/unit/rack/user_spec.rb]: Would probably rephrase to something like "... user has methods with arguments". Then, would use a `let` block and an anonymous class as in other examples.

File: spec/unit/rack/user_spec.rb [hunk 3]
```diff
@@ -189,5 +189,19 @@ def current_user
...
+        def name(hello)
+          "my name is #{hello}"
+        end
+      end
+
+      let(:user_data) { described_class.new(UserRequiringParameters.new).as_json }
+
+      it "is empty" do
```
Comment (xymbol) [inline spec/unit/rack/user_spec.rb]: Would rephrase to make the intention more explicit. For example, "does not call the methods with arguments".

File: lib/airbrake/rack/user.rb [hunk 4]
```diff
@@ -50,7 +50,9 @@ def as_json
...
+                                        @user.method(key).arity == -1)
```
Comment (xymbol) [inline lib/airbrake/rack/user.rb]: Would extract a method and do:

```ruby
def can_call?(sym)
  @user.respond_to?(sym) && @user.method(sym).arity <= 0
end
```

Comment (valdis) [inline lib/airbrake/rack/user.rb]: comparison with <= 0 will not work, because methods with variable number of arguments, that have also required arguments will return < -1

**Comment by valdis** [comment]
third time's the charm :)

File: lib/airbrake/rack/user.rb [hunk 5]
```diff
@@ -50,7 +50,12 @@ def as_json
       private
 
       def try_to_get(key)
-        String(@user.__send__(key)) if @user.respond_to?(key)
+        String(@user.__send__(key)) if can_call_method?(key)
```
Comment (kyrylo) [inline lib/airbrake/rack/user.rb]: My preference would be the following implementation, however I appreciate the feedback from the previous reviewer. ⛳️
 
```rb
def try_to_get(key)
  return unless @user.respond_to?(key)
  return if (-1..0).include?(@user.method(key).arity)
  String(@user.__send__(key))
end

File: spec/unit/rack/user_spec.rb [hunk 6]
```diff
@@ -189,5 +189,21 @@ def current_user
         expect(user_data).to be_empty
       end
     end
+
+    context "when Rack user class has methods with arguments" do
```
Comment (kyrylo) [inline spec/unit/rack/user_spec.rb]: This is a bit cryptic. How about:

> when Rack user defines a field with a parameters

Comment (kyrylo) [inline spec/unit/rack/user_spec.rb]: Corrected my example

File: spec/unit/rack/user_spec.rb [hunk 7]
```diff
@@ -189,5 +189,21 @@ def current_user
...
+        described_class.new(
+          Class.new do
+            def name(hello)
+              "my name is #{hello}"
+            end
+          end.new
+        ).as_json
+      end
+
+      it "does not call the method if it has arguments" do
```
Comment (kyrylo) [inline spec/unit/rack/user_spec.rb]: Continuing from the previous example:

> when Rack user's field expects a parameter
> it skips field invocation

**Comment by kyrylo** [comment]
> note that arity comparison with <= 0 will not work, because methods with variable number of arguments, that have also required arguments will return < -1

It would be nice to have a test for that.

**Comment by xymbol** [comment]
>note that arity comparison with <= 0 will not work, because methods with variable number of arguments, that have also required arguments will return < -1

You're right. Still, I would avoid the double introspection:

```ruby
# using between?
@user.method(name).arity.between? 0, -1

# using include?
[0, 1].include? @user.method(name).arity

# explicit
arity = @user.method(name).arity
arity == 0 || arity == -1
```

Options, options.

File: lib/airbrake/rack/user.rb (hunk 5 above)
Comment (xymbol) [inline lib/airbrake/rack/user.rb]: Didn't see this, suggested something similar below.

Comment (valdis) [inline lib/airbrake/rack/user.rb]: guys, this is not cool, i don't have neither the time nor patience to change the code up time and time again, until all of you like it. Discuss first internally and then give suggestions.

Comment (kyrylo) [inline lib/airbrake/rack/user.rb]: I'm sorry for the confusion @valdis, however I'd like to make it clear that @xymbol doesn't maintain this repo and he's not our employee.

You can see that my badge says "Owner", and @xymbol's doesn't.

don t call methods of user if parameters are required by valdis pull request 755 airbrake airbrake 2017-06-13 16-03-07

Comment (kyrylo) [inline lib/airbrake/rack/user.rb]: @valdis do you want me to take over this PR? Happy to help.

Comment (xymbol) [inline lib/airbrake/rack/user.rb]: @valdis Sorry to read this. My comment was well-intentioned and, it's not a matter of aesthetic taste. Anything that goes into a library of this kind gets called over and over and eventually does affect performance. Peer review is good.

**Comment by kyrylo** [comment]
Forgot about the existence of `#between?`. Cool trick!

**Comment by valdis** [comment]
i hope this is good enough now :)

**Comment by kyrylo** [comment]
Thanks for the gotcha. Given that this is deprecated and overriding those methods isn't super common (first time I see this), I think we can ignore this issue with required params for now.
